src/PostgREST/App.hs

```diff
@@ -9,430 +9,664 @@ Some of its functionality includes:
-      ActionInvoke _     ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-                             ++ rawContentTypes
-                             ++ [CTOpenAPI | tpIsRootSpec target]
-      ActionInspect _    ->  [CTOpenAPI, CTApplicationJSON]
-      ActionInfo         ->  [CTTextCSV]
-      ActionSingleUpsert ->  [CTApplicationJSON, CTSingularJSON, CTTextCSV]
-    serves sProduces cAccepts =
-      case mutuallyAgreeable sProduces cAccepts of
-        Nothing -> Left . errorResponseFor . ContentTypeError . map toMime $ cAccepts
-        Just ct -> Right ct
-
-{-
-  | If raw(binary) output is requested, check that ContentType is one of the admitted rawContentTypes and that
-  | `?select=...` contains only one field other than `*`
--}
-binaryField :: ContentType -> [ContentType] -> Bool -> ReadRequest -> Either Response (Maybe FieldName)
-binaryField ct rawContentTypes isScalarProc readReq
-  | isScalarProc =
-      if ct `elem` rawContentTypes
-        then Right $ Just "pgrst_scalar"
-        else Right Nothing
-  | ct `elem` rawContentTypes =
-      let fieldName = headMay fldNames in
-      if length fldNames == 1 && fieldName /= Just "*"
-        then Right fieldName
-        else Left . errorResponseFor $ BinaryFieldError ct
-  | otherwise = Right Nothing
+postgrestResponse
+  :: AppConfig
+  -> Maybe DbStructure
+  -> SQL.Pool
+  -> UTCTime
+  -> Wai.Request
+  -> Handler IO Wai.Response
+postgrestResponse conf maybeDbStructure pool time req =
+  do
+    body <- lift $ Wai.strictRequestBody req
+
+    dbStructure <-
+      case maybeDbStructure of
+        Just dbStructure ->
+          return dbStructure
+
+        Nothing ->
+          throwError Error.ConnectionLostError
+
+    apiRequest <-
+      liftEither . mapLeft Error.ApiRequestError $
+        Req.userApiRequest
+          (Config.configDbSchemas conf)
+          (Config.configDbRootSpec conf)
+          dbStructure
+          req
+          body
+
+    -- The JWT must be checked before touching the db
+    jwtClaims <- Auth.jwtClaims conf (toS $ Req.iJWT apiRequest) time
+
+    let
+      reqContentTypes = requestContentTypes conf apiRequest
+      acceptContentType = Req.iAccepts apiRequest
+
+    contentType <-
+      case Req.mutuallyAgreeable reqContentTypes acceptContentType of
+        Just ct ->
+          return ct
+
+        Nothing ->
+          throwError . Error.ContentTypeError $ map Types.toMime acceptContentType
+
+    let
+      context apiReq =
+        RequestContext conf dbStructure apiReq contentType
+
+      handleReq apiReq =
+        handleRequest (context apiReq)
+
+    runDbHandler pool (txMode apiRequest) jwtClaims $
+      Middleware.optionalRollback conf apiRequest $
+        Middleware.runPgLocals conf jwtClaims handleReq apiRequest
+
+runDbHandler :: SQL.Pool -> SQL.Mode -> Auth.JWTClaims -> DbHandler a -> Handler IO a
+runDbHandler pool mode jwtClaims handler =
+  do
```

Comment:
This is a regular pattern, not only here: I suggest to move the `do` to the previous line everywhere. This reduces indentation of the whole block by 2 spaces, saving screen real estate.